### PR TITLE
Navigation computer auto offsets

### PR DESCRIPTION
**Pull request by Dennok**
## About The Pull Request

Navigation computer calculate offsets based on connected shuttle
Navigation computer show jump button if find only beacons too

A bit better navigation computer on spawn auto connect.

Refactored `get_containing_dock(atom/A)` to return first found dock.

Removed `z_lock` from shuttle_docker, since it useless in it. Docker jump to ports, not to z lvls

## Why It's Good For The Game

Less calculations for mappers
Fix docking consoles with wrong offsets

## Changelog
:cl:
fix: Fixed docking consoles with wrong offsets
tweak: Navigation computer calculate offsets based on connected shuttle
tweak: Now navigation computer show jump button if find beacon too
/:cl:

### CODE REVIEW

**Review by Timberpoes — CHANGES_REQUESTED** [review]
This completely breaks Nav Computers.

Attempting to use any nav computer will output `The bluespace interference is jamming the console!` - Except when in transit, where it's usable. Completely the opposite of what your changelog suggests.

Can you check on this, fix it if I'm correct and test it?

File: code/modules/shuttle/navigation_computer.dm [hunk 1]
```diff
@@ -102,6 +97,14 @@
 			the_eye.placement_images[I] = list(x_off, y_off)
 
 /obj/machinery/computer/camera_advanced/shuttle_docker/give_eye_control(mob/user)
+	if((locked_traits & ZTRAIT_RESERVED) && !SSmapping.level_trait(z, ZTRAIT_RESERVED))
```
Comment (Dennok) [inline code/modules/shuttle/navigation_computer.dm:100]: If we have reserved level in locked traits, and level has `reserved` trait we block interraction.
```suggestion
	if((locked_traits & ZTRAIT_RESERVED) && SSmapping.level_trait(z, ZTRAIT_RESERVED))
```

File: code/modules/shuttle/navigation_computer.dm [hunk 2]
```diff
@@ -102,6 +97,14 @@
 			the_eye.placement_images[I] = list(x_off, y_off)
 
 /obj/machinery/computer/camera_advanced/shuttle_docker/give_eye_control(mob/user)
+	if((locked_traits & ZTRAIT_RESERVED) && SSmapping.level_trait(z, ZTRAIT_RESERVED))
+		to_chat(user, "<span class='warning'>The bluespace interference is jamming the console!</span>")
+		return
```
Comment (Dennok) [inline code/modules/shuttle/navigation_computer.dm:102]:
```suggestion
```

File: code/modules/shuttle/navigation_computer.dm [hunk 3]
```diff
@@ -102,6 +97,11 @@
 			the_eye.placement_images[I] = list(x_off, y_off)
 
 /obj/machinery/computer/camera_advanced/shuttle_docker/give_eye_control(mob/user)
+
+	var/obj/docking_port/stationary/dock = SSshuttle.get_containing_dock(src)
+	if(dock)
+		jumpto_ports[dock.id] = TRUE
+
```
Comment (Dennok) [inline code/modules/shuttle/navigation_computer.dm:104]:
```suggestion
```

File: code/modules/shuttle/navigation_computer.dm [hunk 4]
```diff
@@ -285,8 +281,16 @@
 
 /obj/machinery/computer/camera_advanced/shuttle_docker/connect_to_shuttle(obj/docking_port/mobile/port, obj/docking_port/stationary/dock, idnum, override=FALSE)
 	if(port && (shuttleId == initial(shuttleId) || override))
+		shuttle_port = port
 		shuttleId = port.id
 		shuttlePortId = "[port.id]_custom"
+
+		//Take info from connected port and calculate amendments
+		var/list/shuttlebounds = shuttle_port.return_coords()
+		view_range = round(max((shuttlebounds[1] - shuttlebounds[3]), (shuttlebounds[2] - shuttlebounds[4]))/2)
```
Comment (Dennok) [inline code/modules/shuttle/navigation_computer.dm:290]:
```suggestion
		view_range = min(round(max(port.width, port.height)*0.5), 15)
```

**Comment by Timberpoes** [comment]
Reviewing your own code, making suggestions and committing suggestions is some utter wizardry that blows my mind.

It also does not fill me with confidence that this is tested.

**Comment by Dennok** [comment]
> Reviewing your own code, making suggestions and committing suggestions is some utter wizardry that blows my mind.

Really this useful thing since it allow me see what i change, and to some extent why. And i use it to code not from home, when i have time.

> It also does not fill me with confidence that this is tested as it's just a string of webedits.

You right, i test it some later.
But as minimum i know that it compiles.

**Comment by Dennok** [comment]
No return to base since it can move eye to centcom level.

**Comment by Dennok** [comment]
Seems like it work. Spawn docker on monastery shuttle and be able park it nearby station.

File: code/modules/shuttle/navigation_computer.dm [hunk 5]
```diff
@@ -26,19 +26,15 @@
...
+			var/obj/docking_port/stationary/S = port
+			if(S.port_destinations == shuttleId)
 				jumpto_ports[S.id] = TRUE
 
-	for(var/V in SSshuttle.stationary)
-		if(!V)
-			continue
-		var/obj/docking_port/stationary/S = V
-		if(jumpto_ports[S.id])
-			z_lock |= S.z
```
Comment (Timberpoes) [inline code/modules/shuttle/navigation_computer.dm:41]: What did this code do and why have you removed it?

File: code/modules/shuttle/navigation_computer.dm [hunk 6]
```diff
@@ -356,8 +359,6 @@
 			stack_trace("SSshuttle.stationary have null entry!")
 			continue
 		var/obj/docking_port/stationary/S = V
-		if(console.z_lock.len && !(S.z in console.z_lock))
-			continue
```
Comment (Timberpoes) [inline code/modules/shuttle/navigation_computer.dm:360]: Same again, what did this code do and why have you removed it?

**Review by Timberpoes — CHANGES_REQUESTED** [review]

File: code/modules/shuttle/navigation_computer.dm (hunk 5 above)
Comment (Dennok) [reply code/modules/shuttle/navigation_computer.dm:41]: This part of code collect to `z_lock` z levels of initial ports available to jump

**Pull request closed without merging**